File: 牛客网/搜索/字符串的排列.java
package 牛客网.搜索;

import java.util.ArrayList;
import java.util.Collections;

/**
 * @Author Aqinn
 * @Date 2021/1/24 2:12 上午
 */
public class 字符串的排列 {

    /**
     * 题目描述:
     * 输入一个字符串,按字典序打印出该字符串中字符的所有排列。
     * 例如输入字符串abc,则按字典序打印出由字符a,b,c所能排列出来的所有字符串abc,acb,bac,bca,cab和cba。
     */

    public static void main(String[] args) {
        字符串的排列 main = new 字符串的排列();
        ArrayList<String> list = main.Permutation("abb");
        for (String str : list) {
            System.out.print(str + " ");
        }
        System.out.println();
    }

    public ArrayList<String> Permutation(String str) {
        ArrayList<String> result = new ArrayList<>();
        if (str.length() == 0)
            return result;
        char[] temp = new char[128];
        for (int i = 0; i < str.length(); i++) {
            temp[str.charAt(i)]++;
        }
        char[] chars = new char[str.length()];
        int count = 0;
        for (int i = 0; i < 128; i++) {
            if (temp[i]-- > 0) {
                chars[count++] = (char) i;
                --i;
            }
        }
        str = String.valueOf(chars);
        recur(str, "", result);
        return result;
    }

    private void recur(String str, String cur, ArrayList<String> result) {
        if (str.length() == 0) {
            if (!result.contains(cur)) {
                result.add(cur);
            }
        } else {
            for (int i = 0; i < str.length(); i++) {
                recur(str.substring(0, i) + str.substring(i + 1), cur + str.charAt(i), result);
            }
        }
    }


}

File: LeetCode/字符串/字符串的排列.kt
package LeetCode.字符串

/**
 * @Author Aqinn
 * @Date 2021/6/22 1:06 下午
 */

/*
题目描述:
输入一个字符串，打印出该字符串中字符的所有排列。
你可以以任意顺序返回这个字符串数组，但里面不能有重复元素。

示例:
输入：s = "abc"
输出：["abc","acb","bac","bca","cab","cba"]

限制：
1 <= s 的长度 <= 8

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/zi-fu-chuan-de-pai-lie-lcof
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */

fun permutation(s: String): Array<String> {
    val arr = s.toCharArray()
    val res = mutableListOf<String>()
    dfs(arr, 0, res)
    return res.toTypedArray()
}

fun dfs(arr: CharArray, idx: Int, res: MutableList<String>) {
    if (idx == arr.size - 1) {
        res.add(String(arr))
        return
    }
    val set = HashSet<Char>()
    for (i in idx until arr.size) {
        if (!set.contains(arr[i])) {
            set.add(arr[i])
            swap(arr, i, idx)
            dfs(arr, idx + 1, res)
            swap(arr, i, idx)
        }
    }
}

fun swap(arr: CharArray, i: Int, j: Int) {
    val temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
}

fun main() {
    println(permutation("abc").contentToString())
}